v7/_posts/design/2018-09-26-$expandForV4.01.md

```diff
@@ -0,0 +1,112 @@
+
+For a 4.0 response, the expanded navigation property suffixed with parentheses MAY be omitted from the select-list if it does not contain a nested $select or $expand, but MUST still be present, without a suffix, if it is explicitly selected.
+```
+
+## Change Summary
+According to the OData spec above, this change is for creating proper expand token in response's context url, corresponding to $expand clause in the request url. Context Url is used as response metadata and is used to control data materialization when response is received by the client.
+We need to make sure that:
+- Context url in the response generated contains correct token for the $expand clause.
+- Client can parse the context url correctly and uses the metadata for response processing.
+
+### V4.01
+
+As required, we are going to add to the context url the parenthesized (either empty or non-empty) navigation property when it is expanded. For example:
+
+Request Url | Context Url in Response
+----------- | -----------------------
+ root/Cities('id')?$expand=TestModel.CapitolCity/Districts                       | root/$metadata#Cities(TestModel.CapitolCity/Districts())/$entity
+ root/Cities('id')?$expand=TestModel.CapitolCity/Districts($select=DistrictName) | root/$metadata#Cities(TestModel.CapitolCity/Districts(DistrictName))/$entity
+ root/Cities('id')?$select=Name&$expand=TestModel.CapitolCity/Districts($select=DistrictName) | root/$metadata#Cities(Name,TestModel.CapitolCity/Districts(DistrictName))/$entity
+ root/Cities('id')?$select=Name,TestModel.CapitolCity/Districts$expand=TestModel.CapitolCity/Districts($select=DistrictName) | root/$metadata#Cities(Name,TestModel.CapitolCity/Districts(DistrictName))/$entity
+
+
+Except for the case (due to known issue https://github.com/OData/odata.net/issues/1104) when a navigation link is both expanded and selected, the parenthesized navigation property will not be present in the context-url, which only has the navigation link in the selected list:
+
+Request Url | Context Url in Response
+----------- | -----------------------
+ root/Cities('id')?$select=**TestModel.CapitolCity/Districts**&$expand=**TestModel.CapitolCity/Districts** | root/$metadata#Cities(**TestModel.CapitolCity/Districts**)/$entity
+ root/Cities('id')?$select=Id,Name,**ExpandedNavProp**&$expand=**ExpandedNavProp**                         | root/$metadata#Cities(Id,Name,**ExpandedNavProp**)/$entity
```

Comment:
>root/$metadata#Cities(Id,Name,**ExpandedNavProp**)/$entity [](start = 109, length = 58)

after code change should be root/$metadata#Cities(Id,Name,ExpandedNavProp,ExpandedNavProp())/$entity